### PR TITLE
Bugfix and improve debug mode

### PR DESCRIPTION
**Pull request by GorogPeter**
-  print parse errors before abort

-  print interpreter error before abort if debug mode set

-  print error at lexer initialization fail

-  print "assert failed" instead of just abort

-  make printing data fancier in debug mode

-  remove dump() in ByteCode and replace it with getName() and printExtra()

-  fix 2 typo

-  bugfix at table access

Old debug mode:
```
function type 0x556f7987a5d0
requiredStackSize 16, requiredStackSizeDueToLocal 0
0: I32Eqz src: 0 dst: 8
16: jump_if_true srcOffset: 8 dst: 64
32: const32 dstOffset: 8 value: 42
48: move64 srcOffset: 8 dstOffset: 0 
64: end resultOffsets: 0  
function type 0x556f7987a5d0
requiredStackSize 32, requiredStackSizeDueToLocal 8
0: I32Eqz src: 0 dst: 16
16: jump_if_true srcOffset: 16 dst: 128
32: const32 dstOffset: 24 value: 1
48: I32Eqsrc1: 0 src2: 24 dst: 16
64: jump_if_true srcOffset: 16 dst: 176
80: const32 dstOffset: 16 value: 7
96: move64 srcOffset: 16 dstOffset: 8 
112: jump dst: 208
128: const32 dstOffset: 16 value: 42
144: move64 srcOffset: 16 dstOffset: 8 
160: jump dst: 208
176: const32 dstOffset: 16 value: 99
192: move64 srcOffset: 16 dstOffset: 8 
208: end resultOffsets: 8  
invoke block1(0) expect value(0) (line: 35) : OK
invoke block1(1) expect value(42) (line: 36) : OK
invoke block2(0) expect value(42) (line: 38) : OK
invoke block2(1) expect value(99) (line: 39) : OK
invoke block2(2) expect value(7) (line: 40) : OK
```

New one:
```
| function type                   : 0x55f83d3bc5d0
| required stack size             :       16 bytes
| required stack size due to local:        0 bytes
| bytecode size                   :      122 bytes
|
|  address        | size | opcode          | note      
|-----------------+------+-----------------+-----------------
|               0 |   24 | I32Eqz          | src: 0 dst: 8
|              24 |   24 | jump_if_true    | srcOffset: 8 dst: 96
|              48 |   24 | const32         | dstOffset: 8 value: 42
|              72 |   24 | move64          | srcOffset: 8 dstOffset: 0 
|              96 |   26 | end             | resultOffsets: 0 
|-----------------+------+-----------------+-----------------


| function type                   : 0x55f83d3bc5d0
| required stack size             :       32 bytes
| required stack size due to local:        8 bytes
| bytecode size                   :      338 bytes
|
|  address        | size | opcode          | note      
|-----------------+------+-----------------+-----------------
|               0 |   24 | I32Eqz          | src: 0 dst: 16
|              24 |   24 | jump_if_true    | srcOffset: 16 dst: 192
|              48 |   24 | const32         | dstOffset: 24 value: 1
|              72 |   24 | I32Eq           | src1: 0 src2: 24 dst: 16
|              96 |   24 | jump_if_true    | srcOffset: 16 dst: 264
|             120 |   24 | const32         | dstOffset: 16 value: 7
|             144 |   24 | move64          | srcOffset: 16 dstOffset: 8 
|             168 |   24 | jump            | dst: 312
|             192 |   24 | const32         | dstOffset: 16 value: 42
|             216 |   24 | move64          | srcOffset: 16 dstOffset: 8 
|             240 |   24 | jump            | dst: 312
|             264 |   24 | const32         | dstOffset: 16 value: 99
|             288 |   24 | move64          | srcOffset: 16 dstOffset: 8 
|             312 |   26 | end             | resultOffsets: 8 
|-----------------+------+-----------------+-----------------

invoke block1(0) expect value(0) (line: 35) : OK
invoke block1(1) expect value(42) (line: 36) : OK
invoke block2(0) expect value(42) (line: 38) : OK
invoke block2(1) expect value(99) (line: 39) : OK
invoke block2(2) expect value(7) (line: 40) : OK
```

### CODE REVIEW

File: src/interpreter/ByteCode.h [hunk 1]
```diff
@@ -256,6 +256,8 @@ class ByteCode {
 
     Opcode opcode() const;
     size_t getSize();
+    virtual const std::string getName() const { return "MISSING NAME"; };
+    virtual void printExtra(size_t pos) { return; };
```
Comment (clover2123) [inline src/interpreter/ByteCode.h:260]: I recommend not to use `virtual` functions in ByteCode because the address of vtable is added for each ByteCode instance which increases the overall ByteCode stream size.

File: src/runtime/Module.cpp [hunk 2]
```diff
@@ -282,7 +282,7 @@ Instance* Module::instantiate(ExecutionState& state, const ExternVector& imports
                          &data);
             }
 
-            if (UNLIKELY(elem->tableIndex() >= numberOfTableTypes() || index >= instance->m_tables[elem->tableIndex()]->size() || index + elem->functionIndex().size() > instance->m_tables[elem->tableIndex()]->size())) {
+            if (UNLIKELY(elem->tableIndex() >= numberOfTableTypes() || index > instance->m_tables[elem->tableIndex()]->size() || index + elem->functionIndex().size() > instance->m_tables[elem->tableIndex()]->size())) {
```
Comment (clover2123) [inline src/runtime/Module.cpp:285]: Why did you change this `>=` operator as to '>' ?

File: src/interpreter/ByteCode.h (hunk 1 above)
Comment (GorogPeter) [reply src/interpreter/ByteCode.h:260]: Okay.
Thank you!

File: src/runtime/Module.cpp (hunk 2 above)
Comment (GorogPeter) [reply src/runtime/Module.cpp:285]: In some cases it causes error (to be more exact, in tests: **binary.wast**, **binary-leb128.wast** and **elem.wast**), when it tries to write nothing to a table that has no size.

Here (from file **elem.wast:174-177**):
```
(module
  (table 0 funcref)
  (elem (i32.const 0))
)
```

Although it didn't cause assert failure, it didn't work propely, I think. I noticed it when I started to print out semantic errors.

**Comment by GorogPeter** [comment]
I closed this pull request to separate the two main improvements (**improving debug mode** and **printing error messages**). Soon I'll make 2 separate pull request for each improvements. Sorry for the overcomplicating.

**Comment by clover2123** [comment]
Ok thanks!
